tidy3d/plugins/autograd/utilities.py

```diff
@@ -158,3 +162,69 @@ def chained(array: NDArray):
         return reduce(lambda x, y: y(x), funcs, array)
 
     return chained
+
+
+def scalar_objective(func: Callable = None, *, has_aux: bool = False) -> Callable:
+    """Decorator to modify a function to return the `.data` attribute of an xarray.DataArray.
+
+    If the function returns an xarray.DataArray, this decorator extracts the `.data` attribute.
+    Additionally, it checks if the return value is scalar and real, raising an exception if it is not.
+    The result is also squeezed to remove single-dimensional entries.
+
+    Parameters
+    ----------
+    func : Callable
+        The function to be decorated.
+    has_aux : bool = False
+        Whether the function returns auxiliary data.
+
+    Returns
+    -------
+    Callable
+        The decorated function.
+
+    Raises
+    ------
+    Tidy3dError
+        If the return value is not scalar or not real.
+    """
+
+    if func is None:
+        return lambda f: scalar_objective(f, has_aux=has_aux)
+
+    @wraps(func)
+    def wrapper(*args: Any, **kwargs: Any) -> Any:
+        result = func(*args, **kwargs)
+        aux_data = None
+
+        # If the result is a tuple, unpack it
```

Comment:
I'm wondering if all of this logic could be handled easier if it was just

```
try:
    result = result.item()
except ValueError as e:
    raise ValueError(...) from e

if not anp.isreal(result):
    raise Tidy3dError("An objective function's return value must be real.")

 return (result, aux_data) if aux_data is not None else result
```

but maybe there's some subtelty here?

Reply:
this won't work for example if the return is a float.. I changed the logic a bit but overall I think most of these checks are necessary if we want to have this work with arbitrary functions

Reply:
ah ok sounds good